x/valset-pref/validator_set.go

```diff
@@ -119,75 +128,143 @@ func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, co
 	if err != nil {
 		return err
 	}
 
-	// the total amount the user wants to undelegate
-	tokenAmt := sdk.NewDec(coin.Amount.Int64())
+	if undelegation.Amount.ToDec().GT(totalDelegatedAmt) {
+		return fmt.Errorf("Total tokenAmountToUndelegate more than delegated amount have %s got %s\n", totalDelegatedAmt, undelegation.Amount)
+	}
 
-	err = k.CheckUndelegateTotalAmount(tokenAmt, existingSet.Preferences)
-	if err != nil {
-		return err
+	// Step 3
+	sort.Slice(valSetRatio, func(i, j int) bool {
+		return valSetRatio[i].VRatio.GT(valSetRatio[j].VRatio)
+	})
+
+	// Step 4
+	totalUnDelAmt, amountToUnDelegate := sdk.NewInt(0), sdk.NewInt(0)
+	if valSetRatio[0].VRatio.LTE(sdk.OneDec()) {
+		fmt.Println("All validators can undelegate normally, falling to happy path exit early")
+		for index, val := range valSetRatio {
+			validator := validators[val.ValAddr.String()]
+
+			// in the last valset iteration we dont calculate it from shares using decimals and trucation,
+			// we use whats remaining to get more accurate value
+			if len(existingSet.Preferences)-1 == index {
+				amountToUnDelegate = undelegation.Amount.Sub(totalUnDelAmt).ToDec().TruncateInt()
+			} else {
+				// Calculate the amount to undelegate based on the existing weights
+				amountToUnDelegate = val.UndelegateAmt
+				totalUnDelAmt = totalUnDelAmt.Add(amountToUnDelegate)
+			}
+
+			sharesAmt, err := validator.SharesFromTokens(amountToUnDelegate)
+			if err != nil {
+				return err
+			}
+
+			_, err = k.stakingKeeper.Undelegate(ctx, delegator, val.ValAddr, sharesAmt) // this has to be shares amount
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
 	}
 
-	// totalDelAmt is the amount that keeps running track of the amount of tokens undelegated
-	totalUnDelAmt := sdk.NewInt(0)
-	amountToUnDelegate := sdk.NewInt(0)
+	// Step 5
+	// `targetRatio`: This is a threshold value that is used to decide how to unbond tokens from validators.
+	// It starts as 1 and is recalculated each time a validator is fully unbonded and removed from the unbonding process.
+	// By reducing the target ratio using the ratio of the removed validator, we adjust the proportions we are aiming for with the remaining validators.
+	targetRatio := sdk.OneDec()
+	amountRemaining := undelegation.Amount
 
-	for i, val := range existingSet.Preferences {
-		valAddr, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
+	// Step 6
+	for len(valSetRatio) > 0 && valSetRatio[0].VRatio.GT(targetRatio) {
+		fmt.Printf("Undelegating fully from validator %s %s\n", valSetRatio[0].ValAddr.String(), valSetRatio[0].DelegatedAmt)
+		_, err = k.stakingKeeper.Undelegate(ctx, delegator, valSetRatio[0].ValAddr, valSetRatio[0].DelegatedAmt.ToDec()) // this has to be shares amount
 		if err != nil {
 			return err
 		}
+		amountRemaining = amountRemaining.Sub(valSetRatio[0].DelegatedAmt)
+		targetRatio = targetRatio.Mul(sdk.OneDec().Sub(valSetRatio[0].Weight))
```

Comment:
Don't quite follow exactly how we do re-calculation and use target ratio here , can you give me some examples with numbers? 👀

Reply:
yup so if we have 2 validators and have userA who has staked 20token
```
ValA = 15token, weight: 0.5
ValB = 5Token, weight: 0.5

Run undelegate 20token

ValA VRatio =  0.5 * 20.0 = 10, VRatio = 10/15 = 0.6667
ValB VRatio =  0.5 * 20.0 = 10, VRatio = 10/5 = 2.0

determine the maxVRatio = 2.0, we set targetVRatio to 1, since we want to normalize all the weights 

For validator B, VRatio 2.0 > 1.0. So, 
- perform full undelegation, 
- reduce amount remaining (20 - stake of B, which is 5) = 15, 
- adjust target ratio (1.0 * (1 - RatioInValset of B, which is 0.5)) = 0.5, 
- Remove B from validators list.
- do this until valSetRatio is LTE target ratio

Now only Validator A is left and amount remaining is 15, so we undelegate the rest from ValA

```

this is a full flow of how the algorithm works, please lmk if that makes sense. Happy to explain more :))

Reply:
Not sure if this is valid, I don't think it is currently possible to have a vratio greater than 1. Would love to see a test case that touches steps 5-7

Reply:
The issue with the described calculation in this example is the Vratio. In this example we take 10/15 and 10/5. This would be possible if we used GetValSetPreferences method in the withdraw logic, but we instead use GetValSetPreferencesWithDelegations. This transforms the weights based on the amount actually delegated to the validators. So in that case, the numerator could never be larger than the denominator.